File: src/modules/06-expiry-control/expiry-guard.ts
import type { CapAlert } from '../../types/cap.js';
import { getLogger } from '../../utils/logger.js';
import { capTiming } from '../01-cap-ingestion/cap-parser.js';

const logger = getLogger();

/**
 * Expiry-aware submission control — requirement #6.
 *
 * Before submitting each batch to the SMSC, check the CAP alert's real
 * `expires` timestamp; the moment it is reached, halt all further submission.
 *
 * The guard is seeded with the real CAP timing output of module 01
 * (`capTiming(alert).expiresAt`) — never a placeholder date. A configurable
 * lead margin protects against clock skew between TURANT and the SMSC, and
 * `EXPIRY_HALT_SUBMISSION` gates the hard halt.
 */

export interface ExpiryGuardOptions {
  /** Real CAP expires timestamp (null = no expiry declared). */
  expiresAt: Date | null;
  /** Injectable clock for deterministic tests (defaults to `new Date()`). */
  now?: () => Date;
  /** Safety margin (ms) applied before the declared expiry. Default 0. */
  leadMarginMs?: number;
  /** `EXPIRY_HALT_SUBMISSION` — when false, submission is not halted. */
  haltEnabled?: boolean;
}

export interface ExpiryStatus {
  canSubmit: boolean;
  reason: 'ok' | 'expired' | 'no-expiry' | 'halt-disabled';
  remainingMs: number | null;
  expiresAt: Date | null;
}

/**
 * Real expiry logic. `canSubmit()` is safe to call inside a submission loop
 * before every batch — it is cheap and recomputes from the current instant.
 */
export class ExpiryGuard {
  private readonly expiresAt: Date | null;
  private readonly now: () => Date;
  private readonly leadMarginMs: number;
  private readonly haltEnabled: boolean;

  constructor(options: ExpiryGuardOptions) {
    this.expiresAt = options.expiresAt;
    this.now = options.now ?? (() => new Date());
    this.leadMarginMs = options.leadMarginMs ?? 0;
    this.haltEnabled = options.haltEnabled ?? true;
  }

  /** True when submission should proceed for the current instant. */
  canSubmit(): boolean {
    return this.status().canSubmit;
  }

  /** Detailed status for logging / counting expired messages. */
  status(): ExpiryStatus {
    if (!this.haltEnabled) {
      return { canSubmit: true, reason: 'halt-disabled', remainingMs: null, expiresAt: this.expiresAt };
    }
    if (!this.expiresAt) {
      return { canSubmit: true, reason: 'no-expiry', remainingMs: null, expiresAt: null };
    }
    const nowMs = this.now().getTime();
    const expiresMs = this.expiresAt.getTime();
    const remainingMs = expiresMs - nowMs - this.leadMarginMs;
    if (remainingMs <= 0) {
      return { canSubmit: false, reason: 'expired', remainingMs, expiresAt: this.expiresAt };
    }
    return { canSubmit: true, reason: 'ok', remainingMs, expiresAt: this.expiresAt };
  }

  /** Milliseconds until the (margin-adjusted) expiry; null when no expiry/halt. */
  remainingMs(): number | null {
    return this.status().remainingMs;
  }

  /**
   * Mark t5 on the shared trace when submission halts at expiry (one of the
   * two legitimate t5 conditions — the other is all expected DLRs in).
   */
  async markExpiryTrace(traceKey: string): Promise<void> {
    const { traceStore } = await import('../../tracing/trace-store.js');
    logger.warn({ traceKey }, 'submit.halt_expired');
    await traceStore.mark(traceKey, 't5', 'alert.expiry', Date.now());
  }
}

/** Build a guard from a real parsed CAP alert (module 01 output). */
export function expiryGuardForAlert(alert: CapAlert, overrides: Omit<ExpiryGuardOptions, 'expiresAt'> = {}): ExpiryGuard {
  return new ExpiryGuard({ expiresAt: capTiming(alert).expiresAt, ...overrides });
}

/** One-shot check for the current instant, with default halt behavior. */
export function isExpiredNow(expiresAt: Date | null, leadMarginMs = 0): boolean {
  return !new ExpiryGuard({ expiresAt, leadMarginMs }).canSubmit();
}

File: src/main/java/com/turant/expiry/ExpiryGuard.java
package com.turant.expiry;

import com.turant.types.cap.CapAlert;
import com.turant.types.cap.CapTiming;
import com.turant.cap.CapParser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.function.Supplier;

/**
 * Expiry-aware submission control - requirement #6.
 * 
 * Before submitting each batch to the SMSC, check the CAP alert's real
 * `expires` timestamp; the moment it is reached, halt all further submission.
 * 
 * The guard is seeded with the real CAP timing output of module 01
 * (`capTiming(alert).expiresAt`) - never a placeholder date.
 * 
 * NOT a Spring component - instances are created per-alert using factory methods.
 * 
 * Migrated from TypeScript Module 06 expiry-guard.ts
 */
public class ExpiryGuard {
    
    private static final Logger logger = LoggerFactory.getLogger(ExpiryGuard.class);
    
    private final Instant expiresAt;
    private final Supplier<Instant> clock;
    private final long leadMarginMs;
    private final boolean haltEnabled;
    
    /**
     * Create expiry guard with options.
     */
    public ExpiryGuard(ExpiryGuardOptions options) {
        this.expiresAt = options.expiresAt;
        this.clock = options.clock != null ? options.clock : Instant::now;
        this.leadMarginMs = options.leadMarginMs != null ? options.leadMarginMs : 0;
        this.haltEnabled = options.haltEnabled != null ? options.haltEnabled : true;
    }
    
    /**
     * Create expiry guard from CAP alert.
     */
    public static ExpiryGuard forAlert(CapAlert alert, CapParser capParser, ExpiryGuardOptions overrides) {
        CapTiming timing = capParser.parseCapTiming(alert.info());
        
        ExpiryGuardOptions options = new ExpiryGuardOptions();
        options.expiresAt = timing.expiresAt();
        
        if (overrides != null) {
            if (overrides.clock != null) options.clock = overrides.clock;
            if (overrides.leadMarginMs != null) options.leadMarginMs = overrides.leadMarginMs;
            if (overrides.haltEnabled != null) options.haltEnabled = overrides.haltEnabled;
        }
        
        return new ExpiryGuard(options);
    }
    
    /**
     * True when submission should proceed for the current instant.
     */
    public boolean canSubmit() {
        return status().canSubmit;
    }
    
    /**
     * Detailed status for logging / counting expired messages.
     */
    public ExpiryStatus status() {
        if (!haltEnabled) {
            return new ExpiryStatus(true, "halt-disabled", null, expiresAt);
        }
        
        if (expiresAt == null) {
            return new ExpiryStatus(true, "no-expiry", null, null);
        }
        
        Instant now = clock.get();
        long nowMs = now.toEpochMilli();
        long expiresMs = expiresAt.toEpochMilli();
        long remainingMs = expiresMs - nowMs - leadMarginMs;
        
        if (remainingMs <= 0) {
            return new ExpiryStatus(false, "expired", remainingMs, expiresAt);
        }
        
        return new ExpiryStatus(true, "ok", remainingMs, expiresAt);
    }
    
    /**
     * Milliseconds until the (margin-adjusted) expiry; null when no expiry/halt.
     */
    public Long remainingMs() {
        return status().remainingMs;
    }
    
    /**
     * Mark expiry trace when submission halts at expiry.
     */
    public void markExpiryTrace(String traceKey) {
        logger.warn("Submission halted due to expiry: traceKey={}", traceKey);
        
        // TODO: Mark t5 on trace store when implemented
        // traceStore.mark(traceKey, "t5", "alert.expiry", System.currentTimeMillis());
    }
    
    /**
     * One-shot check for the current instant, with default halt behavior.
     */
    public static boolean isExpiredNow(Instant expiresAt, long leadMarginMs) {
        ExpiryGuardOptions options = new ExpiryGuardOptions();
        options.expiresAt = expiresAt;
        options.leadMarginMs = leadMarginMs;
        
        return !new ExpiryGuard(options).canSubmit();
    }
    
    /**
     * Options for creating an expiry guard.
     */
    public static class ExpiryGuardOptions {
        public Instant expiresAt;
        public Supplier<Instant> clock;
        public Long leadMarginMs;
        public Boolean haltEnabled;
    }
    
    /**
     * Status of expiry check.
     */
    public record ExpiryStatus(
        boolean canSubmit,
        String reason,
        Long remainingMs,
        Instant expiresAt
    ) {}
}
